### ISSUE TITLE
Crash on Ubuntu 22.04 

### ISSUE THREAD

**Issue by KeplerC** [ros2/rmw_cyclonedds#391]
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 22.04 
- Installation type:
  - binary
- Version or commit hash:
  - 1.3.1
> $ ros2 doctor --report
> ...
> rmw_cyclonedds_cpp                        : latest=1.3.3, local=1.3.3

- DDS implementation: rmw_cyclonedds_cpp
- > $ ros2 doctor --report
> ...
>    RMW MIDDLEWARE
> middleware name    : rmw_cyclonedds_cpp

#### Steps to reproduce issue

Minimally working dockerfile: 
```Dockerfile
ARG DISTRO=rolling
FROM ros:$DISTRO
ARG DISTRO
RUN apt update && apt install -y curl gnupg2 lsb-release && \
    curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key -o /usr/share/keyrings/ros-archive-keyring.gpg
    
RUN echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/ros-archive-keyring.gpg] http://packages.ros.org/ros2/ubuntu jammy main" | sudo tee /etc/apt/sources.list.d/ros2.list > /dev/null
RUN apt install -y ros-rolling-desktop ros-rolling-rmw-cyclonedds-cpp

ENTRYPOINT [ "/ros_entrypoint.sh" ]
CMD ["bash"]
```

and run 
```
$  . /opt/ros/rolling/setup.bash
$ export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp
$ ros2 run demo_nodes_cpp talker
# or 
$ ros2 run demo_nodes_cpp listener
```

#### Expected behavior
listener and talker communicate with each other. 

#### Actual behavior

```
$ ros2 run demo_nodes_cpp talker
$ ros2 run demo_nodes_cpp listener
[ERROR] [1651185922.565368862] [rclcpp]: executor taking a loaned message from topic '/parameter_events' unexpectedly failed: Loaning is not supported, at ./src/rmw_node.cpp:3345, at ./src/rcl/subscription.c:335
[ERROR] [1651185922.565616782] [rclcpp]: executor taking a loaned message from topic '/parameter_events' unexpectedly failed: Loaning is not supported, at ./src/rmw_node.cpp:3345, at ./src/rcl/subscription.c:335
[ERROR] [1651185922.565662817] [rclcpp]: executor taking a loaned message from topic '/parameter_events' unexpectedly failed: Loaning is not supported, at ./src/rmw_node.cpp:3345, at ./src/rcl/subscription.c:335
```
or sometimes they end up with segmentation fault or stack smashing. 
![image](https://user-images.githubusercontent.com/25167951/165860023-89fdf6c6-019d-44ea-b36f-f8acec80e4d8.png)

**Comment by KeplerC** [comment]
If unset RMW_IMPLEMENTATION=rmw_cyclonedds_cpp, the talker and listener example works perfectly.

**Comment by KeplerC** [comment]
Hi Thanks for the clarification.  I tried to further debug my own code and found the issue: 
The cyclonedds retrieved by `apt-get` is not compatible with the `ros:rolling` docker image. In my dockerfile, I did not put `apt update` before `apt get`, and the cyclone is retrieved from an incompatible build. This problem is resolved by changing image `ros:rolling` to `ubuntu:jammy`. 

I am not 100% certain why this happens. I am closing this issue since the problem is solved, and I am happy to reopen if further investigation is neede.d
